axelrod/strategies/axelrod_second.py

```diff
@@ -199,3 +202,194 @@ def strategy(self, opponent: Player) -> Action:
+        """
+        if opponent.history[-1] == D: 
+            self.opponent_consecutive_defections += 1
+        else:
+            self.opponent_consecutive_defections = 0
+
+        if self.num_turns_after_good_defection == 2:
+            self.num_turns_after_good_defection = 0
+            self.two_turns_after_good_defection_ratio = (
+                ((self.two_turns_after_good_defection_ratio 
+                * self.two_turns_after_good_defection_ratio_count) 
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - ((self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]]))) 
+                / (self.two_turns_after_good_defection_ratio_count + 1)
+                )
+            self.two_turns_after_good_defection_ratio_count += 1
+        elif self.num_turns_after_good_defection == 1:
+            self.num_turns_after_good_defection = 2
+            self.one_turn_after_good_defection_ratio = (
+                ((self.one_turn_after_good_defection_ratio 
+                * self.one_turn_after_good_defection_ratio_count)
+                + (3 - (3 * self.dict[opponent.history[-1]])) 
+                + (2 * self.dict[self.history[-1]]) 
+                - (self.dict[opponent.history[-1]] 
+                * self.dict[self.history[-1]])) 
+                / (self.one_turn_after_good_defection_ratio_count + 1)
+                )
+            self.one_turn_after_good_defection_ratio_count += 1
+   
+    def strategy(self, opponent: Player) -> Action:
+
+        if len(self.history) == 0:
+            return C
+
+        self.current_score = compute_final_score(zip(self.history, opponent.history))
+       
+        self.update_cooperative_state(opponent)
+        if  self.num_turns_after_good_defection in [1, 2]:
+            return C        
+        
+        if (self.current_score[0] / ((len(self.history)) + 1)) >= 2.25:
+            probability = (
+                (.95 - (((self.one_turn_after_good_defection_ratio)
+                + (self.two_turns_after_good_defection_ratio) - 5) / 15)) 
+                + (1 / (((len(self.history))+1) ** 2))
+                - (self.dict[opponent.history[-1]] / 4)
+                )
+            if random.random() <= probability: 
+                return C
+            self.num_turns_after_good_defection = 1
+            return D
+        elif (self.current_score[0] / ((len(self.history)) + 1)) >= 1.75:
+            probability = (
+                (.25 + ((opponent.cooperations + 1) / ((len(self.history)) + 1)))
+                - (self.opponent_consecutive_defections * .25) 
+                + ((self.current_score[0] 
+                - self.current_score[1]) / 100) 
+                + (4 / ((len(self.history)) + 1))
+                )
+            if random.random() <= probability:
+                return C
+            self.num_turns_after_good_defection = 1
```

Comment:
This should not be here. As you can see here: https://github.com/Axelrod-Python/TourExec/blob/v0.3.0/src/strategies/k67r.f#L38 we don't change state if this random test fails.

Reply:
This will require unit tests to be fixed.